File: android/app/src/main/java/app/farmsy/android/ui/ProductImage.kt
package app.farmsy.android.ui

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.ui.theme.FarmsyColors

/// A product photograph from res/drawable-nodpi, by the slug the server names —
/// the Android twin of iOS ProductImage.swift.
///
/// The files are the 512px WebPs from farmsy-web/design/product-images as
/// product_<slug>.webp (resource names cannot hold a hyphen), so a new image is
/// a file copy and nothing else. Falls back to the emoji the tile used to show,
/// so a slug the app does not ship (a product added server-side before the
/// next release) still reads as something.
@Composable
fun ProductImage(slug: String?, fallback: String, size: Dp, corner: Dp = 12.dp) {
    val context = LocalContext.current
    val id = slug?.let { ProductImageIds.resolve(context, it) } ?: 0
    val shape = RoundedCornerShape(corner)
    if (id != 0) {
        Image(painterResource(id), null, Modifier.size(size).clip(shape), contentScale = ContentScale.Crop)
    } else {
        Box(Modifier.size(size).clip(shape).background(FarmsyColors.creamFill), contentAlignment = Alignment.Center) {
            Text(fallback, fontSize = (size.value * 0.5f).sp)
        }
    }
}

/// Resource ids by slug, looked up once: getIdentifier is a reflective string
/// search and the same eggs appear on Home, Shopping and Discover in one scroll.
object ProductImageIds {
    private val cache = HashMap<String, Int>()

    fun resolve(context: android.content.Context, slug: String): Int = cache.getOrPut(slug) {
        val name = "product_" + slug.replace('-', '_')
        context.resources.getIdentifier(name, "drawable", context.packageName)
    }
}

File: Farmsy/App/ProductImage.swift
import SwiftUI

/// A product photograph from the bundle, by the slug the server names.
///
/// The files are the 512px WebPs from farmsy-web/design/product-images,
/// dropped into Farmsy/ProductImages as product-<slug>.webp (Xcode flattens
/// them into the bundle root, hence the prefix), so a new image
/// is a file copy and nothing else. Decoded once and cached: the same eggs
/// appear on Home, Shopping and Discover in one scroll.
///
/// Falls back to the emoji the tile used to show, so a slug the bundle does
/// not have (a product added server-side before the next release) still
/// reads as something.
struct ProductImage: View {
    let slug: String?
    /// A second file to try before the emoji: a recipe picture that is not
    /// bundled yet shows its product's photograph.
    var fallbackSlug: String? = nil
    var fallback: String = "🌱"
    var size: CGFloat = 56
    var corner: CGFloat = Radius.thumb

    var body: some View {
        Group {
            if let image = slug.flatMap(ProductImageCache.image) ?? fallbackSlug.flatMap(ProductImageCache.image) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFill()
            } else {
                Text(fallback)
                    .font(.system(size: size * 0.5))
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .background(Color.creamFill)
            }
        }
        .frame(width: size, height: size)
        .clipShape(RoundedRectangle(cornerRadius: corner, style: .continuous))
    }
}

enum ProductImageCache {
    private static var cache: [String: UIImage] = [:]
    private static var missing: Set<String> = []

    /// A seasonal product that is the same thing as a shopping item shares its
    /// photograph. The server sends the mapping as `image`; this is the same
    /// table (farmsy-web src/lib/productImages.ts) for builds that run against
    /// a server that does not yet.
    static let alias: [String: String] = [
        "aardbei": "strawberry", "appel": "apples", "peer": "pears", "asperge": "asparagus",
        "pompoen": "pumpkin", "tomaat": "tomatoes", "ui": "onions", "wortel": "carrot",
        "paddenstoel": "mushrooms", "aardappel": "potatoes", "honing": "honey",
        "eieren": "eggs", "kaas": "cheese",
    ]

    static func image(_ raw: String) -> UIImage? {
        let slug = alias[raw] ?? raw
        if let hit = cache[slug] { return hit }
        if missing.contains(slug) { return nil }
        guard let url = Bundle.main.url(forResource: "product-" + slug, withExtension: "webp"),
              let img = UIImage(contentsOfFile: url.path)
        else { missing.insert(slug); return nil }
        cache[slug] = img
        return img
    }
}
